src/osd/scrubber/pg_scrubber.cc

```diff
@@ -1033,13 +1203,77 @@ void PgScrubber::on_init()
   m_pg->publish_stats_to_osd();
 }
 
-/*
- * Note: as on_replica_init() is likely to be called twice (entering
- * both ReplicaWaitUpdates & ActiveReplica), its operations should be
- * idempotent.
- * Now that it includes some state-changing operations, we need to check
- * m_active against double-activation.
- */
+
+void PgScrubber::on_repl_reservation_failure()
+{
+  const seconds penalty_period =
+      seconds{m_pg->get_cct()->_conf.get_val<int64_t>(
+	  "osd_scrub_busy_replicas_penalty")};
+  dout(10) << fmt::format(
+		  "{}: penalty period: {}s", __func__, penalty_period.count())
+	   << dendl;
+
+  if (m_scrub_job->on_reservation_failure(
+	  penalty_period, std::move(*m_active_target))) {
+
+
+    utime_t now_is = m_scrub_queue.scrub_clock_now();
+    m_depenalize_timer = std::make_unique<depenalize_timer_t>(
+	m_osds, this, now_is + utime_t{penalty_period});
+  }
+
+  m_active_target.reset();
+  clear_pgscrub_state();
+}
+
+PgScrubber::depenalize_timer_t::depenalize_timer_t(
+    OSDService* osds,
+    PgScrubber* scrbr,
+    utime_t timeout)
+    : m_osds{osds}
+    , m_scrbr{*scrbr}
+{
+  m_depenalize_cb = new LambdaContext(
+      [this, scrbr = scrbr, pgid = scrbr->m_pg_id]([[maybe_unused]] int r) {
+	if (!m_depenalize_cb) {
+	  // already cancelled
+	  return;
+	}
+	m_depenalize_cb = nullptr;
```

Comment:
Is this not a data race with the pg threads?

Reply:
Larger question, why isn't penalize a special case of not_before?

Reply:
For the seconds comment: see my general answer re the penalized.